qui/tray/devices.py

```diff
@@ -256,56 +232,84 @@ def __init__(self, menu: Gtk.Menu, device_manager):
             self.positions[dev_type] = pos
             self.separators[dev_type] = separator
 
-        self.device_manager.connect_to_signal("Added", self.add)
-        self.device_manager.connect_to_signal("Removed", self.remove)
-
-    def add(self, dev_obj_path: dbus.ObjectPath):
-        dev = self.device_manager[dev_obj_path]
-        if dev['dev_class'] not in DEV_TYPES:
+        for devclass in DEV_TYPES:
+            self.dispatcher.add_handler('device-attach:' + devclass,
+                                        self.device_attached)
+            self.dispatcher.add_handler('device-detach:' + devclass,
+                                        self.device_detached)
+            self.dispatcher.add_handler('device-list-change:' + devclass,
+                                        self.device_change)
+
+    def update_device_list(self):
+        devices = {}
+        for vm in self.qapp.domains:
+            for devclass in DEV_TYPES:
+                for device in vm.devices[devclass]:
+                    devices[device] = []
+
+        for vm in self.qapp.domains:
+            for devclass in DEV_TYPES:
+                for device in vm.devices[devclass].attached():
+                    if device in devices:
+                        # occassionally ghost UnknownDevices appear when a
+                        # device was removed but not detached from a VM
+                        devices[device].append(vm)
+
+        for device in [dev for dev in devices if dev not in self.menu_items.keys()]:
+            self.add(device, devices[device])
+
+        for device in [dev for dev in self.menu_items.keys() if dev not in devices]:
+            self.remove(device)
+
+    def device_change(self, _vm, _event, **_kwargs):
+        self.update_device_list()
```

Comment:
You get vm of which device list changed - using this you could avoid iterating over all the VMs.